dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/Java8IncompatibleReferenceCheck.java

```diff
@@ -0,0 +1,145 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.opensource.dependencies;
+
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+
+import com.google.cloud.tools.opensource.classpath.ClassFile;
+import com.google.cloud.tools.opensource.classpath.ClassPathBuilder;
+import com.google.cloud.tools.opensource.classpath.ClassPathEntry;
+import com.google.cloud.tools.opensource.classpath.ClassPathResult;
+import com.google.cloud.tools.opensource.classpath.LinkageChecker;
+import com.google.cloud.tools.opensource.classpath.LinkageProblem;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSetMultimap;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.logging.Logger;
+import org.eclipse.aether.artifact.Artifact;
+
+/**
+ * A tool to find Java 8-incompatible references in the class files in a BOM. It checks the class
+ * files in the following artifacts:
+ *
+ * <ul>
+ *   <li>The artifacts in the {@code dependencyManagement} section of the BOM
+ *   <li>The artifacts in the dependency graph for each of the artifact above
+ *   <li>The latest version of the artifacts in the dependency graphs
+ * </ul>
+ *
+ * <p>If it finds references to classes in the core library, which are not present in Java 8, it
+ * exits with status code 1; otherwise it exits with status code 0.
+ *
+ * @see <a
+ *     href="https://github.com/GoogleCloudPlatform/cloud-opensource-java/wiki/Java-8-incompatible-references-of-java.nio.Buffer-classes-generated-by-Java-9--compilers"
+ *     >Java 8 incompatible references of java.nio.Buffer classes generated by Java 9 compilers</a>
+ */
+public class Java8IncompatibleReferenceCheck {
+
+  private static final Logger logger =
+      Logger.getLogger(Java8IncompatibleReferenceCheck.class.getName());
+
+  public static void main(String[] arguments)
+      throws MavenRepositoryException, IOException, URISyntaxException {
+
+    if (arguments.length != 1) {
+      System.err.println("Specify a path to the BOM file");
+      System.exit(1);
+    }
+
+    URI exclusionFileUri =
+        Java8IncompatibleReferenceCheck.class
+            .getClassLoader()
+            .getResource("java8-incompatible-reference-check-exclusion.xml")
+            .toURI();
+    Path exclusionFile = Paths.get(exclusionFileUri).toAbsolutePath();
```

Comment:
I'm surprised this works since it's a resource in a jar, not a file. Could we skip both URIs and paths here and get a stream from the resource instead?

Reply:
This works because it reads the XML file from `target` directory. With maven-exec-plugin (used by java8-incompatible-reference-check.sh), the value (toString) of `exclusionFile` is `Users/suztomo/cloud-opensource-java/dependencies/target/classes/java8-incompatible-reference-check-exclusion.xml`, not from a JAR file.

`LinkageChecker.create()` takes `Path` for an exclusion file.

Reply:
I'll trust you that it works, but this still looks very brittle to me. Can we add a method to LinkageChecker that takes an InputStream to avoid this.

Reply:
I tried adding another factory method for LinkageChecker to take URL (not InputStream, because ExcludedErrors already takes URL) but the change is unexpectedly large https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/1625/commits/01b6dc4023dec3417224af53fdc530d96c0f63e7

I will just use Path for the exclusion xml as second parameter, without URL or InputStream.

Reply:
Now the exclusion file is passed as the second parameter. No need to deal with URL.